### PR TITLE
DOC: Several documentation fixes (broken links, incorrect sphinx syntax, duplicate sections)

### PR DESCRIPTION
**Pull request by MSeifert04**
The PR links PyObject references in narrative documentation and also fixes "invalid section header" warning in "linalg.multi_dot" docstring.

It seems weird that `:c:type:'PyObject *'` can't be resolved as link while `:c:type:'PyObject'` works. 😕 Is there something wrong in the intersphinx?

(This time with `[ci skip]`)

### CODE REVIEW

**Comment by eric-wieser** [comment]
Guessing by `:c:func:` you mean `:c:type:`?

**Comment by MSeifert04** [comment]
you're right, ... corrected.

**Comment by MSeifert04** [comment]
I'm not sure about the workflow but is it okay to combine several changes (like the one about the `invalid section header`-warning from numpydoc) and the one fixing (or rather changing) the `PyObject *`-links? And is the commit message alright?

File: doc/source/user/c-info.beyond-basics.rst [hunk 1]
```diff
@@ -173,7 +173,7 @@ each array is for PyArray_ITER_NEXT to be called for each of the inputs. This
 incrementing is automatically performed by
 :c:func:`PyArray_MultiIter_NEXT` ( ``obj`` ) macro (which can handle a
 multiterator ``obj`` as either a :c:type:`PyArrayMultiObject *` or a
```
Comment (eric-wieser) [inline doc/source/user/c-info.beyond-basics.rst]: Does `PyArrayMultiObject *` work just fine here then?

File: numpy/linalg/linalg.py [hunk 2]
```diff
@@ -2357,9 +2357,7 @@ def multi_dot(arrays): return functools.reduce(np.dot, arrays)
     >>> # or
     >>> A.dot(B).dot(C).dot(D)
 
-
-    Example: multiplication costs of different parenthesizations
-    ------------------------------------------------------------
+    Another example: multiplication costs of different parenthesizations.
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Not really an example at all - perhaps notes, or more details or something. Perhaps a third-level heading (`~~~`) under notes instead?

**Comment by eric-wieser** [comment]
> 'm not sure about the workflow but is it okay to combine several changes

File: doc/source/user/c-info.beyond-basics.rst (hunk 1 above)
Comment (MSeifert04) [inline doc/source/user/c-info.beyond-basics.rst]: Haha! That's funny because there is no `PyArrayMultiObject` defined so any link (with or without `*`) couldn't work. 😄 

However from my understanding `:c:type::`-declared variables can have the `*` and link correctly. See https://docs.scipy.org/doc/numpy/reference/c-api.array.html#data-type-checking for working `PyArrayObject *`-link.

File: numpy/linalg/linalg.py (hunk 2 above)
Comment (MSeifert04) [inline numpy/linalg/linalg.py]: I can also put it in a `Notes`-section :)

**Comment by MSeifert04** [comment]
So if I had 10 other documentation-related fixes then it would be fine to amend them? Or will that be hellish to review?

**Comment by eric-wieser** [comment]
One Pr might be ok, but separate commits might be sensible

**Comment by eric-wieser** [comment]
Regarding `PyObject *` - I'd rather we fixed something like that upstream, rather than work around it and forget.

**Comment by MSeifert04** [comment]
@eric-wieser I don't think that's so easy to fix upstream. I'm not sure but there is something weird going on with name-mangling in numpydoc. That's a lot beyond my current capabilities.

I actually added a lot of additional "link" fixes. I grouped them by the kind of change (should be easy to review commit-by-commit), just let me know if I need to squash them :)

**Comment by eric-wieser** [comment]
> That's a bit (lot) beyond my current capabilities.

Yeah, I tryed to debug this and was struggling too. My only concern is that `PyObject *` is possibly more useful at a glance, and realistically after this change, we'll never switch back if sphinx ever fixes it.

At some point I might try and produce a minimal testcase to file an issue against sphinx

File: doc/source/reference/c-api.array.rst [hunk 3]
```diff
@@ -204,7 +204,7 @@ From scratch
     new array is *descr*.
 
     If *subtype* is of an array subclass instead of the base
-    :c:data:`&PyArray_Type`, then *obj* is the object to pass to
+    :c:data:`&PyArray_Type<PyArray_Type>`, then *obj* is the object to pass to
```
Comment (eric-wieser) [inline doc/source/reference/c-api.array.rst]: What happened here before?

File: doc/source/reference/c-api.array.rst [hunk 4]
```diff
@@ -1545,7 +1545,7 @@ Conversion
 
 .. c:function:: PyObject* PyArray_GetField(PyArrayObject* self, PyArray_Descr* dtype, int offset)
 
-    Equivalent to :meth:`ndarray.getfield` (*self*, *dtype*, *offset*). Return
+    Equivalent to :meth:`ndarray.getfield<numpy.ndarray.getfield>` (*self*, *dtype*, *offset*). Return
```
Comment (eric-wieser) [inline doc/source/reference/c-api.array.rst]: Putting `~numpy.` before here doesn't work then?

**Review by eric-wieser — COMMENTED** [review]
Great set of fixes - thanks. Few nits.

File: doc/source/reference/c-api.array.rst (hunk 3 above)
Comment (MSeifert04) [inline doc/source/reference/c-api.array.rst]: It just wasn't rendered as link: https://docs.scipy.org/doc/numpy/reference/c-api.array.html#c.PyArray_NewFromDescr

File: doc/source/reference/c-api.array.rst (hunk 4 above)
Comment (MSeifert04) [inline doc/source/reference/c-api.array.rst]: It would work! But it would render as `getfield` instead of `ndarray.getfield`.

**Comment by MSeifert04** [comment]
I addessed some additional issues. However I don't want to make this a "work in progress" so I'll stop now 😄

File: doc/source/reference/c-api.types-and-structures.rst [hunk 5]
```diff
@@ -6,7 +6,7 @@ Python Types and C-Structures
 
 Several new types are defined in the C-code. Most of these are
 accessible from Python, but a few are not exposed due to their limited
-use. Every new Python type has an associated :c:type:`PyObject *` with an
+use. Every new Python type has an associated :c:type:`PyObject *<PyObject>` with an
```
Comment (charris) [inline doc/source/reference/c-api.types-and-structures.rst]: I'd rewrite that as "has a pointer to an associated :c:type:`PyObject` ,,,"? However, I think this whole section is somewhat lacking, what we usually get is a `PyTypeObject` that subclasses the `PyObject` type. I think it would be better to quote something from the Python documentation.

Comment (MSeifert04) [inline doc/source/reference/c-api.types-and-structures.rst]: > However, I think this whole section is somewhat lacking

I thought this passage is about the difference between `PyLongType` and `PyLongObject`. So to say every python type (long type) has an associated object (long object). I don't quite understand what you mean about every type has a pointer to an associated object?

But I tried to only fix broken links and incorrect attempts to link. Not okay?

File: doc/source/reference/internals.code-explanations.rst [hunk 6]
```diff
@@ -460,12 +460,12 @@ Ufuncs allow other array-like classes to be passed seamlessly through
 the interface in that inputs of a particular class will induce the
 outputs to be of that same class. The mechanism by which this works is
 the following. If any of the inputs are not ndarrays and define the
-:obj:`__array_wrap__` method, then the class with the largest
-:obj:`__array_priority__` attribute determines the type of all the
+:obj:`~numpy.class.__array_wrap__` method, then the class with the largest
```
Comment (eric-wieser) [inline doc/source/reference/internals.code-explanations.rst]: What is `class` referring to here? Is this some weird sphinx feature, or a quick hack? What does this end up linking to?

**Comment by eric-wieser** [comment]
> But I tried to only fix broken links and incorrect attempts to link. Not okay?

This mostly looks good to me. Thanks for keeping `PyObject *`and just fixing the links

File: doc/source/reference/internals.code-explanations.rst (hunk 6 above)
Comment (MSeifert04) [inline doc/source/reference/internals.code-explanations.rst]: No, that is actually called `class` in the numpy documentation on subclassing: [`numpy.class.__array_wrap__`](https://docs.scipy.org/doc/numpy-dev/reference/arrays.classes.html#numpy.class.__array_wrap__)

Comment (eric-wieser) [inline doc/source/reference/internals.code-explanations.rst]: Ok, that's pretty weird! Still, I guess if this PR is all about fixing links, then this is the right thing to do

**Review by eric-wieser — APPROVED** [review]
I'm sure there's more to do with fixing docs, but as an exercise in fixing links, this looks great.

File: doc/source/reference/c-api.ufunc.rst [hunk 7]
```diff
@@ -124,9 +122,7 @@ Functions
     :param unused:
         Unused and present for backwards compatibility of the C-API.
 
-.. c:function:: PyObject* PyUFunc_FromFuncAndDataAndSignature(PyUFuncGenericFunction* func,
-   void** data, char* types, int ntypes, int nin, int nout, int identity,
-   char* name, char* doc, int unused, char *signature)
```
Comment (eric-wieser) [inline doc/source/reference/c-api.ufunc.rst]: Line wrapping is not possible then? Backslashes at EOL?

File: numpy/linalg/linalg.py [hunk 8]
```diff
@@ -2357,14 +2357,13 @@ def multi_dot(arrays): return functools.reduce(np.dot, arrays)
     >>> # or
     >>> A.dot(B).dot(C).dot(D)
 
-
-    Example: multiplication costs of different parenthesizations
-    ------------------------------------------------------------
-
+    Notes
```
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: We already have a notes section higher up.

Also, the changes to this file kinda feel different to all the rest - maybe put them in their own PR?

Comment (MSeifert04) [inline numpy/linalg/linalg.py]: maybe I'm blind but I can't find the other `Notes`-section.

File: doc/source/reference/c-api.ufunc.rst (hunk 7 above)
Comment (MSeifert04) [inline doc/source/reference/c-api.ufunc.rst]: I haven't checked but [`c-api.iterator`](https://github.com/numpy/numpy/blame/master/doc/source/reference/c-api.iterator.rst#L285) also contains longish lines. I can see what happens with backslashes. Is there a line-length-limit for `.rst` files?

Comment (MSeifert04) [inline doc/source/reference/c-api.ufunc.rst]: Backslashes seem to work, should I use them instead?

File: numpy/linalg/linalg.py (hunk 8 above)
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: No, it is I who am blind...

File: doc/source/reference/c-api.ufunc.rst (hunk 7 above)
Comment (eric-wieser) [inline doc/source/reference/c-api.ufunc.rst]: Yes please!

File: numpy/linalg/linalg.py (hunk 8 above)
Comment (MSeifert04) [inline numpy/linalg/linalg.py]: I will open another PR for this change then :)

Comment (eric-wieser) [inline numpy/linalg/linalg.py]: Feel free to leave this in, since it's correct. Just add the backslashes on the lines you unwrapped, and this'll be good to merge

File: doc/source/reference/c-api.types-and-structures.rst (hunk 5 above)
Comment (charris) [inline doc/source/reference/c-api.types-and-structures.rst]: A generic `PyObject` is a C `struct` that begins with a `PyObject` (i.e., subclasses PyObject), which in turn is a struct that contains a reference count and a pointer to a `PyObjectType` struct. I don't know how that fits with the rest of the paragraph 8-)

**Comment by MSeifert04** [comment]
I used EOL-backslashes so the function signature lines wrap at roughly 80 characters.

File: doc/source/user/c-info.ufunc-tutorial.rst [hunk 9]
```diff
@@ -1058,9 +1058,7 @@ What follows is the full specification of PyUFunc_FromFuncAndData, which
 automatically generates a ufunc from a C function with the correct signature.
 
 
-.. c:function:: PyObject *PyUFunc_FromFuncAndData( PyUFuncGenericFunction* func,
-   void** data, char* types, int ntypes, int nin, int nout, int identity,
-   char* name, char* doc, int unused)
+.. c:function:: PyObject *PyUFunc_FromFuncAndData( PyUFuncGenericFunction* func, void** data, char* types, int ntypes, int nin, int nout, int identity, char* name, char* doc, int unused)
```
Comment (eric-wieser) [inline doc/source/user/c-info.ufunc-tutorial.rst]: You missed the one that started this conversation!

Comment (MSeifert04) [inline doc/source/user/c-info.ufunc-tutorial.rst]: 😅 fixed

**Comment by MSeifert04** [comment]
sorry, that last change really bloated the "number of lines changed" count.

**Comment by MSeifert04** [comment]
Anything else that needs to be adressed?

**Comment by eric-wieser** [comment]
If no one else chimes in, I'll merge this tomorrow

**Comment by charris** [comment]
@eric-wieser Merge?

**Comment by eric-wieser** [comment]
Thanks @MSeifert04!

**Comment by MSeifert04** [comment]
Thanks for reviewing and merging it 👍 

I just noticed that the PR has no "component: documentation" tag, I don't know how you do bookkeeping here but maybe it was simply forgotten?
